pkg/upgrade/upgrades/first_upgrade_test.go

```diff
@@ -296,3 +356,87 @@ func TestFirstUpgradeRepair(t *testing.T) {
 	tdb.CheckQueryResults(t, "SELECT * FROM test.public.foo", [][]string{{"1", "2", "42"}})
 	tdb.CheckQueryResults(t, "SELECT test.public.f()", [][]string{{"1"}})
 }
+
+// TestFirstUpgradeRepairBatchSize validates that the maximum batch size
+// used by the repair upgrade is sustainable.
+func TestFirstUpgradeRepairBatchSize(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	// This test is fairly slow because of the large number of
+	// descriptors.
+	skip.UnderDuress(t)
+
+	var (
+		v0 = clusterversion.MinSupported.Version()
+		v1 = clusterversion.Latest.Version()
+	)
+
+	ctx := context.Background()
+	settings := cluster.MakeTestingClusterSettingsWithVersions(v1, v0, false /* initializeVersion */)
+	// We intentionally create contention on the system.descriptor table by having
+	// a high priority txn that intentionally updates rows in this table. This txn
+	// also holds a lease on the system database descriptor, which we will wait to
+	// be released. Reducing the lease duration makes this part of the test speed
+	// up.
+	lease.LeaseDuration.Override(ctx, &settings.SV, time.Second*30)
+	require.NoError(t, clusterversion.Initialize(ctx, v0, &settings.SV))
+	testServer, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
+		Settings: settings,
+		Knobs: base.TestingKnobs{
+			Server: &server.TestingKnobs{
+				DisableAutomaticVersionUpgrade: make(chan struct{}),
+				ClusterVersionOverride:         v0,
+			},
+			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+		},
+	})
+	defer testServer.Stopper().Stop(ctx)
+
+	const totalDescriptorsToTest = 5000
```

Comment:
this is a lot of descriptors! nice!

are we concerned about adding this slow of a test to CI (idk how slow this one is compared to others)? would this make sense as a roachtest?